### PR TITLE
[Bots] Expanded Bot Spell Settings List.

### PR DESCRIPTION
**Pull request by Aeadoin**
Implemented new Bot Command "**^enforcespelllist**" this command will force the targeted Bot to ONLY cast spells from their existing Spell Settings List.

"**^enforcespelllist**" being set to True will cause no entries to display with "**^spelllist**", to add new spells to the Spell Settings List:

1. Set "**^enforcespelllist**" to false.
2. Run ""**^spelllist**"
3. Add new spells to your Spell Settings List.
4. When completed adding spells, set "**^enforcespelllist**" to true.

Adds Say Link Functionality to Bot Commands "**^spelllist**" and "**^spellsettingslist**"

These Say Links allow you to quickly Add/Remove Spells to/from your Bots Spell Settings List, along with Toggle the Enable/Disabled State of Spells. You can also click the Spell Name to get the Spell Effect Description to a dialog box.

image

"**^spelllist**" can now be used with an optional min level parameter ""**^spelllist minlevel**"

Cleaned up existing messages (can still view priority options with **^spellsettingsupdate** command

Fixed potential crash issues with AI_AddBotSpells() with where it was being loaded in (inside constructor, but some database calls would not have been completed at time of function call).

### CODE REVIEW

File: zone/bot.h [hunk 1]
```diff
@@ -789,6 +793,7 @@ class Bot : public NPC {
 	bool _pauseAI;
 	uint8 _stopMeleeLevel;
 	int m_expansion_bitmask;
+	bool _enforcespellsettings;
```
Comment (Kinglykrab) [inline zone/bot.h:796]: I know the old ones don't follow the same standard, but we try to prefix member variables with `m_` like `m_enforce_spell_settings`.

File: zone/bot_command.cpp [hunk 2]
```diff
@@ -10242,7 +10259,7 @@ void bot_command_spell_list(Client* c, const Seperator *sep)
 		c->Message(
 			Chat::White,
 			fmt::format(
-				"Usage: {} [Spell ID] [Priority] [Min Level] [Max Level] [Min HP] [Max HP]",
+				"Usage: {} Optional Min Level 0 - 255",
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10262]: `"Usage: {} [Min Level] (Level is optional)" maybe? Just a bit confusing.

File: zone/bot_command.cpp [hunk 3]
```diff
@@ -10255,7 +10272,13 @@ void bot_command_spell_list(Client* c, const Seperator *sep)
 		return;
 	}
 
-	my_bot->ListBotSpells();
+	uint8 minlevel = 0;
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10275]: `min_level`

File: zone/bot_command.cpp [hunk 4]
```diff
@@ -10255,7 +10272,13 @@ void bot_command_spell_list(Client* c, const Seperator *sep)
...
+
+	if (sep->IsNumber(1)) {
+		minlevel = atoi(sep->arg[1]);
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10278]: `min_level = static_cast<uint8>(std::stoul(sep->arg[1]));`

File: zone/bot_command.cpp [hunk 5]
```diff
@@ -10698,12 +10706,94 @@ void bot_command_spell_settings_update(Client *c, const Seperator *sep)
...
+		(std::stoi(sep->arg[1]) ? true : false) :
+		atobool(sep->arg[1])
+	);
+
+	my_bot->SetBotEnforceSpellSetting(toggle, true);
+
+	c->Message(
+		Chat::White,
+		fmt::format(
+			"{}'s Spell Settings List entries are now {}",
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10792]: Period.

File: zone/bot.h [hunk 6]
```diff
@@ -608,6 +611,7 @@ class Bot : public NPC {
 	void ListBotSpellSettings();
 	void LoadBotSpellSettings();
 	bool UpdateBotSpellSetting(uint16 spell_id, BotSpellSetting* bs);
+	void SetBotEnforceSpellSetting(bool enforcespellsettings, bool save);
```
Comment (Kinglykrab) [inline zone/bot.h:614]: `bool save = false` so it's default and you don't have to send `false` in `SetBotEnforceSpellSetting`.

File: zone/bot_command.cpp [hunk 7]
```diff
@@ -6683,6 +6686,20 @@ void bot_subcommand_bot_spawn(Client *c, const Seperator *sep)
...
 	}
 
+	my_bot->GetBotOwnerDataBuckets();
+	my_bot->GetBotDataBuckets();
+	my_bot->LoadBotSpellSettings();
+	if (!my_bot->AI_AddBotSpells(my_bot->GetBotSpellID())) {
+		c->Message(
+			Chat::White,
+			fmt::format(
+				"Failed to load spells for '{}' (ID {})",
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:6696]: Period.

File: zone/bot.cpp [hunk 8]
```diff
@@ -10757,22 +10770,15 @@ void Bot::ListBotSpellSettings()
...
-				bs.second.priority,
-				GetLevelString(bs.second.min_level, bs.second.max_level),
-				GetHPString(bs.second.min_hp, bs.second.max_hp)
+				Saylink::Silent(fmt::format("^botspellidinfo {}", bs.first), spells[bs.first].name),
+				Saylink::Silent(
+					fmt::format("^spellsettingstoggle {} {}",
+					bs.first, bs.second.is_enabled ? "False" : "True"),
+					bs.second.is_enabled ? "Enabled" : "Disabled"
+				),
+				Saylink::Silent(fmt::format("^spellsettingsdelete {}", bs.first), "Yes")
```
Comment (Kinglykrab) [inline zone/bot.cpp:10781]: Would prefer saylinks are "Add", "Enable", "Disable", similar to this.
image

File: zone/bot.cpp [hunk 9]
```diff
@@ -10697,28 +10713,25 @@ void Bot::ListBotSpells()
...
-			Chat::White,
-			fmt::format(
-				"Spell {} | Priority: {} Health: {}",
-				spell_number,
-				s.priority,
-				GetHPString(s.min_hp, s.max_hp)
-			).c_str()
-		);
+		auto b = bot_spell_settings.find(s.spellid);
+		if(b == bot_spell_settings.end() && s.minlevel >= minlevel) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:10717]: `if (`

File: zone/bot.cpp [hunk 10]
```diff
@@ -10697,28 +10713,25 @@ void Bot::ListBotSpells()
...
+			bot_owner->Message(
+				Chat::White,
+				fmt::format(
+					"Spell {} | Spell: {} Add Spell: {}",
```
Comment (Kinglykrab) [inline zone/bot.cpp:10721]: Saylink comment below.

File: zone/bot.cpp [hunk 11]
```diff
@@ -10697,28 +10713,25 @@ void Bot::ListBotSpells()
...
+					spell_number,
+					Saylink::Silent(
+						fmt::format("^botspellidinfo {}", s.spellid),
+						spells[s.spellid].name
+					),
+					Saylink::Silent(
+						fmt::format("^spellsettingsadd {} {} {} {}", s.spellid, s.priority, s.min_hp, s.max_hp), "Yes")
```
Comment (Kinglykrab) [inline zone/bot.cpp:10728]: Same thing, all saylinks.

File: zone/bot.h [hunk 12]
```diff
@@ -597,7 +600,7 @@ class Bot : public NPC {
 	int GetExpansionBitmask();
 	void SetExpansionBitmask(int expansion_bitmask, bool save = true);
 
-	void ListBotSpells();
+	void ListBotSpells(uint8 minlevel);
```
Comment (Kinglykrab) [inline zone/bot.h:603]: `min_level`

File: zone/bot.h [hunk 13]
```diff
@@ -315,6 +315,8 @@ class Bot : public NPC {
 		int16 iManaCost,
 		int32 iRecastDelay,
 		int16 iResistAdjust,
+		uint8 minlevel,
+		uint8 maxlevel,
```
Comment (Kinglykrab) [inline zone/bot.h:319]: `min_level` and `max_level`

File: zone/bot_command.cpp [hunk 14]
```diff
@@ -10698,12 +10706,94 @@ void bot_command_spell_settings_update(Client *c, const Seperator *sep)
...
+	auto my_bot = ActionableBots::AsTarget_ByBot(c);
+	if (!my_bot) {
+		c->Message(Chat::White, "You must target a bot that you own to use this command.");
+		return;
+	}
+
+	bool toggle = (
+		sep->IsNumber(1) ?
+		(std::stoi(sep->arg[1]) ? true : false) :
+		atobool(sep->arg[1])
```
Comment (Kinglykrab) [inline zone/bot_command.cpp:10784]: `Strings::ToBool(sep->arg[1])`

**Comment by Kinglykrab** [comment]
Beautiful work, small comments. Otherwise looks great.

File: zone/bot.cpp [hunk 15]
```diff
@@ -10757,22 +10768,15 @@ void Bot::ListBotSpellSettings()
...
-				bs.first,
-				bs.second.is_enabled ? "Yes" : "No"
-			).c_str()
-		);
-
-		bot_owner->Message(
-			Chat::White,
-			fmt::format(
-				"Setting {} | Priority: {} Levels: {} Health: {}",
+				"Setting: {} | Spell: {} | State: {} | {}",
```
Comment (Kinglykrab) [inline zone/bot.cpp:10771]: `Setting {}`

**Review by Kinglykrab — APPROVED** [review]

**Comment by MortimerGreenwald** [comment]
This is some really cool stuff. TY can't wait to test it.
-Mort

**Pull request merged**